content/post/graceful-k8s-delpoyments/index.md

```diff
@@ -0,0 +1,205 @@
+---
+
+*[Photo by Kammeran Gonzalez-Keola](https://www.pexels.com/@kammeran-gonzalez-keola-3137381) used under [Pexels License](https://www.pexels.com/license/)*
+
+At Rakuten we default to Kubernetes to run our services, and k8s enables us to operate robust systems at scale. However, it is not perfect and in edge cases k8s may behave in unexpected ways. For example, during rolling deployments we expect no request failures - a reasonable expectation, right? But when we ran functional tests during a deployment we notice requests failing - to our surprise. After researching this for a while we learned that...
+
+> K8s will send traffic to terminating pods.
+
+Even after starting the pod shutdown process, which includes sending a `TERM` signal to pod, k8s still sends new requests to the pod. Because there is no orchestration between k8s sending `TERM` signal and removing the pod from its service endpoint list. Those 2 operations can happen in any order, possibly with delay in between them. Yikes 😬.
+
+In this article, we explore the Kubernetes pod eviction mechanism and show different approaches to avoid these surprising errors in apps deployed with k8s.
+
+## Kubernetes Pod Termination Process
+
+A pod can be evicted for multiple reasons. For example, when a node is drained, via `kubectl delete`, or when the scheduler evicts pods to allow the execution of higher-priority ones. The eviction process starts when the API server modifies the state of a pod in `etcd` to the `Terminating` state. The node’s Kubelet and the endpoints-controller are continuously watching the pod's state. Once they notice the termination state they start the eviction process:
+
+1. Kubelet performs the pod eviction.
+1. The endpoints-controller handles the endpoint removal process.
+
+Both operations are asynchronous.
+
+![SIGTERM in Kubernetes](./images/sigterm.png)
+
+When kubelet acknowledges that a pod should be terminated, it initiates a shutdown sequence for each container in the pod. First, it runs [the container’s pre-stop hook](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#hook-handler-execution) (if it exists), then sends a `TERM` signal, and finally waits for the termination of the container. This sequential process should take less than 30 seconds (or the value in seconds specified in the [`spec.terminationGracePeriodSeconds`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.21/#podspec-v1-core) field). If the container still running beyond this time, kubelet waits for 2 more seconds, and then kills the container forcibly by sending a `KILL` signal.
+
+![SIGKILL in Kubernetes](./images/sigkill.png)
+
+In parallel, the endpoints-controller removes the pod’s endpoint by requesting the API server. This server notifies all the kube-proxies on the worker nodes. Each kube-proxy removes the endpoint from the iptables rules in its node. Note that changes in the iptables do not affect already established connections.
```

Comment:
I thought BPF now replaced iptables in all modern distros?

Reply:
tbh I'm not sure and didn't find a clear answer by quick search. I'm asking one of my devops guys to confirm.

Reply:
So yes eBPF is set to replace iptables in k8s and container networking solutions, so we have to caveat that statement there, I'll add a footnote for it. (For the established connections point - I don't know if eBPF makes a difference, we'd have to try).

The overall issue of routing state propagating through the whole cluster is _probably_ still there, I don't see how eBPF could make distributed state more consistent (just CAP wise) - but I don't know enough to give a qualified statement really.

Reply:
The iptables usage we are specifically discussing is by kube-proxy. It looks like kube-proxy is not going to dump iptables but rather that modern Kubernetes cluseters can choose to dump kube-proxy itself instead. The replacement is Cilium:
https://archive.fosdem.org/2020/schedule/event/replacing_iptables_with_ebpf/attachments/slides/3622/export/events/attachments/replacing_iptables_with_ebpf/slides/3622/Cilium_FOSDEM_2020.pdf

Also:
https://cilium.io/blog/2019/08/20/cilium-16#kubeproxy-removal

The end-note should that the behavior mentioned here also applies to Cilium (or at least seems to apply to it), but there may be more caveats that you have to keep in mind. For instance, Cilium seems to have [a bug](https://github.com/cilium/cilium/issues/14844) that causes **existing connections** to also fail during pod termination, making graceful shutdown harder.
See also the [original issue](https://github.com/cilium/cilium/issues/10217).

Reply:
I put a footnote on this part and added a closing note outlining that this is still a problem for eBPF CNI implementations. Let me know what you think of that (or if it's wrong^^)